coreneuron/utils/reports/report_configuration_parser.cpp

```diff
@@ -0,0 +1,118 @@
+   THE POSSIBILITY OF SUCH DAMAGE.
+   */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <iostream>
+#include <string>
+#include <vector>
+#include <string.h>
+#include "coreneuron/nrniv/nrn_assert.h"
+#include "coreneuron/utils/reports/nrnreport.h"
+#include "coreneuron/utils/tokenizer.hpp"
+#include "coreneuron/nrnoc/mech_mapping.hpp"
+//#include "coreneuron/nrnoc/membfunc.h"
+
+#define MAX_LINE_LENGTH 1024
+//extern "C" int nrn_get_mechtype    (const char*);
+  
+  /*
+   * ---           ---
+   *     Internals
+   * ---           ---
+   */
+
+/*
+ * Split filter string ("mech.var_name") into mech_id and var_name
+ */
+  void parse_filter_string (const char* filter, ReportConfiguration& config) {
+    std::vector<std::string> tok (tokenize(filter, '.'));
+    strncpy(config.mech_name, tok[0].c_str(), tok[0].size());
+    strncpy(config.var_name,  tok[1].c_str(), tok[1].size());
+  }
+  /*
+   * ---                           ---
+   *     Public API implementation
+   * ---                           ---
+   */
+
+std::vector<ReportConfiguration> create_report_configurations (const char* conf_file, const char* output_dir) {
+    std::vector<ReportConfiguration> reports;
+    int num_reports = 0;
+    char report_on [MAX_REPORT_NAME_LEN]= "";
+    char raw_line [MAX_LINE_LENGTH] = "";
+    int is_soma;
+    int* gids;
+    FILE *fp = fopen(conf_file, "r");
+    if (!fp) {
+      printf("Error while reading %s \n", conf_file);
+      return reports;
+    }
+    fgets(raw_line, MAX_LINE_LENGTH, fp);
+    sscanf(raw_line, "%d\n", &num_reports);
+    for(int i = 0; i < num_reports; i++) {
+      reports.push_back(ReportConfiguration());
+      ReportConfiguration& report = reports[reports.size()-1];
+      report.mech_id          = -1; // to be filled after initialization of simulation, cannot be deduced before
+      fgets(raw_line, MAX_LINE_LENGTH, fp);
+      sscanf(raw_line, "\n%s %s %s %s %s %s %d %lf %lf %lf %d\n",
+          report.name,
+          report.target_name,
+          report.type_str,
+          report_on,
+          report.unit,
+          report.format,
+          &is_soma,
+          &report.report_dt,
+          &report.start,
+          &report.stop,
+          &report.num_gids);
+      for (int i = 0; i < MAX_REPORT_NAME_LEN; i++) {
+        report.type_str[i] = tolower(report.type_str[i]);
+      }
+      sprintf(report.output_path, "%s/%s", output_dir, report.name);
+
+      if (! strcmp (report.type_str, "compartment") && (! is_soma)) report.type = CompartmentReport;
+      if (! strcmp (report.type_str, "compartment") && ( is_soma))  report.type = SomaReport;
+      if (! strcmp (report.type_str, "synapse"))                    report.type = SynapseReport;
+      if (report.type == SynapseReport)
+        parse_filter_string(report_on, report);
+      if(report.num_gids) {
+        gids = (int*) calloc(report.num_gids, sizeof(int));
+        fread(gids, sizeof(int), report.num_gids, fp);
```

Comment:
why not directly in into vector?

Reply:
it is an std::set to get quick find algorithm when used. There is no bulk insertion or range constructor like that in API of std::set for c++ < 11